### ISSUE TITLE
ASN lookups 

### ISSUE THREAD

**Issue by BobyMCbobs** [kubernetes-sigs/oci-proxy#20]
# Background
Using collected+collated ASN data from aggregators, the published IP range information from the [cloud credits](https://www.cncf.io/credits) vendors, the source IP can be resolved all the way up to the provider's name and [publisher-verified ASNs](https://github.com/kubernetes/k8s.io/tree/c954d7e2ea28bc3d2c524d72590e8753410da3d9/registry.k8s.io/infra/meta/asns).

Currently this data is being used effectively by the [_public-log-asn-matcher_ job found in k8s.io/image/public-log-asn-matcher](https://github.com/kubernetes/k8s.io/tree/main/images/public-log-asn-matcher), which provides the BigQuery dataset for a detailed DataStudio report to display the (PII) Kubernetes public artifact traffic information, it uses:
- ASN data
  - https://github.com/kubernetes/k8s.io/blob/97379d1e1b03265af62269063b35fe8114f4d81d/images/public-log-asn-matcher/pg-init.d/lib-01-prepare-company-asn-data.sh
- vendor IP ranges
  - https://github.com/kubernetes/k8s.io/blob/155fea4000adc9362b85c8b89b68480abcc1dac8/images/public-log-asn-matcher/pg-init.d/lib-04-load-into-a-bigquery-dataset.sh#L51-L67
- publisher-verified ASNs
  - https://github.com/kubernetes/k8s.io/blob/155fea4000adc9362b85c8b89b68480abcc1dac8/images/public-log-asn-matcher/pg-init.d/lib-04-load-into-a-bigquery-dataset.sh#L31-L49

This rich data will provide everything needed to redirect the traffic to the closest provider which will save costs.

# Implementation
There are a few potential ways this could be implemented, but I think the easiest path should be:
- pull in separately pushed & rendered data, like that is produced from 
  https://github.com/kubernetes/k8s.io/blob/97379d1e1b03265af62269063b35fe8114f4d81d/images/public-log-asn-matcher/pg-init.d/lib-01-prepare-company-asn-data.sh#L30 and 
  https://github.com/kubernetes/k8s.io/blob/97379d1e1b03265af62269063b35fe8114f4d81d/images/public-log-asn-matcher/pg-init.d/lib-01-prepare-company-asn-data.sh#L42
  (flatten the data out even more?)
  on a schedule as a go-routine
- load the data into memory after changes
- make simple look ups against it (very rough pseudo code as example)
```go
// IP range cidr -> providerName
type Lookup map[string]string
lookup := getLookups()
// from https://github.com/kubernetes/k8s.io/tree/main/registry.k8s.io/infra/meta/asns
verifiedASNMetadata := getVerifiedASNMetadata()
// defaults to k8s.gcr.io
registryProvider := getDefaultRegistryProvider()
sourceIP := req.Header.Get("X-Real-Ip")
ipRanges := getIPRanges()
for _, ipRange := range ipRanges {
  _, subnet, _ := net.ParseCIDR(ipRange)
  if !subnet.Contains(sourceIP) {
    continue
  }
  providerName, ok := lookup[ipRange]
  if !ok {
    continue
  }
  registryProvider = verifiedASNMetadata[providerName].RedirectsTo.Registry
  break
}
http.Redirect(w, r, "https://"+registryProvider+path, http.StatusPermanentRedirect)
```
- the registryProvider backends should also be pinged periodically in the background to ensure no client errors after redirection

**Comment by aojea** [comment]
> * load the data into memory after changes

do you have an estimation on how much memory it needs?

**Comment by ameukam** [comment]
> sourceIP := req.Header.Get("X-Real-Ip")

Are we collecting this or keep it in the log ?

**Comment by BobyMCbobs** [comment]
> > * load the data into memory after changes
> 
> do you have an estimation on how much memory it needs?

@aojea, unsure yet on memory usage but expect it to be minimal with the PR

**Comment by BobyMCbobs** [comment]
> > sourceIP := req.Header.Get("X-Real-Ip")
> 
> Are we collecting this or keep it in the log ?

@ameukam, in order to redirect based on IP this is how the real source IP would be gotten. Unsure if it's the same header in Cloud Run

**Comment by aojea** [comment]
> @aojea, unsure yet on memory usage but expect it to be minimal with the PR

a full Internet rib dump is about 1G https://data.ris.ripe.net/rrc00/

**Comment by BobyMCbobs** [comment]
> > @aojea, unsure yet on memory usage but expect it to be minimal with the PR
> 
> [a full Internet rib dump ](https://www.ripe.net/analyse/internet-measurements/routing-information-service-ris/ris-raw-data)is about 1G https://data.ris.ripe.net/rrc00/

@aojea, this data will be pre-processed and ready to be picked up by the service as a plain CSV file

**Comment by dims** [comment]
@BobyMCbobs please avoid a goroutine that runs in-process of the oci-proxy, let's split that into a separate go program that processes the data and spits out a file.

the oci-proxy can watch using inotify on the file and update the in-memory lookup table when the file is refreshed.

**Comment by aojea** [comment]
> @aojea, this data will be pre-processed and ready to be picked up by the service as a plain CSV file

yeah, but a binary compressed that will be several factors greater in csv, where are you going to store the data, are you going to query it directly from the file?

**Comment by BobyMCbobs** [comment]
> @BobyMCbobs please avoid a goroutine that runs in-process of the oci-proxy, let's split that into a separate go program that processes the data and spits out a file.

How does Google Cloud Run go with running multiple containers together and sharing data? I haven't played around with it.

> the oci-proxy can watch using inotify on the file and update the in-memory lookup table when the file is refreshed.

Thank you for your advice @dims, this sounds good.

**Comment by BobyMCbobs** [comment]
> > @aojea, this data will be pre-processed and ready to be picked up by the service as a plain CSV file
> 
> yeah, but a binary compressed will be several factors greater in csv, where are you going to store the data, are you going to query it directly from the file?

The output data for the file will just be a map of IP cidr to provider name after all of the ASN processing, so it shouldn't be overly large
```go
// IP range cidr -> providerName
type Lookup map[string]string
```

@aojea, i'd suggest such a file be committed to a repo for easy fetching and store it on local ephemeral disk in the container (then load into memory).

**Comment by BenTheElder** [comment]
I don't think we should take any of the approaches mentioned so far.

1) We currently have no indication that there will be providers beyond AWS, please let's focus on the immediate cost reduction and not add unnecessary complexity.

2) We may need to be region aware when redirecting, which is not something we can do with ASN data.

3) IP addresses allocated to AWS are very infrequently updated, there's no reason to be watching these at runtime, this is just an additional cost + failure mode at runtime.

I have an implementation that reads AWS published IP range info and bakes it into the binary at development time. I am waiting for our decision on 2) before polishing and PRing.

**Comment by dims** [comment]
ok cool. let's stop iterating here then @BobyMCbobs

**Comment by dims** [comment]
@BenTheElder let's carve off the number 2. into a separate issue and assign to @jaypipes please

**Comment by BenTheElder** [comment]
Also noting: Naive golang CIDR checking even with pre-parsing is *substantially* slower than e.g. a compressed trie (which is what Linux uses for routing). I lost the exact numbers but I have benchmarks and it was at least multiple orders of magnitude off the top of my head even with a pretty generic off the shelf implementation. The performance considerations are a bit different if we need to match by region and we will need provider specific data for regions.

**Comment by BobyMCbobs** [comment]
> I don't think we should take any of the approaches mentioned so far.
> 
> 1. We currently have no indication that there will be providers beyond AWS

Why do you say that?

> please let's focus on the immediate cost reduction and not add unnecessary complexity.

I believe this solution will help with cost reduction and is the right way to go.

> 2. We may need to be region aware when redirecting, which is not something we can do with ASN data.

as noted in the desciption, in public-log-vendor-matcher, IP ranges are pulled in from each vendor
https://github.com/kubernetes/k8s.io/blob/155fea4000adc9362b85c8b89b68480abcc1dac8/images/public-log-asn-matcher/pg-init.d/lib-04-load-into-a-bigquery-dataset.sh#L51-L67
this will be combined with the ASN data in a meaningful way

> 3. IP addresses allocated to AWS are very infrequently updated, there's no reason to be watching these at runtime, this is just an additional cost + failure mode at runtime.

@BenTheElder, I'm sure we can find some way to implement it that works well.

**Comment by dims** [comment]
@BobyMCbobs we should be laser focused on cost-reduction first. so i agree with @BenTheElder

**Comment by BenTheElder** [comment]
> Why do you say that?

In ~5 years of infra migration we do not have any resources of this type from vendors beyond GCP and AWS*, and we're still determining the AWS bits. 

There are many things that will need to be considered before we scale out to additional providers (consider e.g. the behavior of the image promoter and failure with pushing to N>2 providers, when any one provider fails?), and we should not allow those to impede the initial cost reduction we are pursuing. We've discussed this in quite a bit more detail in the design doc and in the SIG K8s Infra meetings.

> I believe this solution will help with cost reduction and is the right way to go.

I don't see how this helps with the AWS+GCP hosting. We have agreed with AWS to route AWS users to a copy they host for us, not any other user, and we have no agreements with any other providers. We need to roll out this cost reduction ASAP to avoid exceeding our GCP budget. AIUI at this time we're not pursuing additional image hosts until at least the costs are under control.

> > We may need to be region aware when redirecting, which is not something we can do with ASN data.
>
> as noted in the desciption, in public-log-vendor-matcher, IP ranges are pulled in from each vendor
https://github.com/kubernetes/k8s.io/blob/155fea4000adc9362b85c8b89b68480abcc1dac8/images/public-log-asn-matcher/pg-init.d/lib-04-load-into-a-bigquery-dataset.sh#L51-L67
this will be combined with the ASN data in a meaningful way

That doesn't address the comment, AWS publishes data that includes IP address ranges by the region in AWS (see https://github.com/kubernetes-sigs/oci-proxy/issues/22), we cannot detect this from ASN data, as the regions will be in the same ASN. We may need to do this to keep costs down.

> > 3. IP addresses allocated to AWS are very infrequently updated, there's no reason to be watching these at runtime, this is just an additional cost + failure mode at runtime.
>
> @BenTheElder, I'm sure we can find some way to implement it that works well.

The point is, there's no real upside to having runtime machinery for pulling down IP address data. Cloud run means we expect very fast application spin up/down and processing this data at runtime is an unnecessary cost in addition to adding a new failure mode to an application that needs to be extremely cheap + reliable. If the redirector is not maximally cheap and reliable then we will not meet our goals, we need to minimize being a source of new infra costs and we need to ensure stability for rapid roll out. Every additional runtime branch we don't strictly need is an unnecessary risk, in addition to the wasted compute / memory.

It is also not important that we 100% accurately detect every new AWS IP range as soon as they are added, we only need to redirect the majority of AWS-origin requests to the AWS origin hosting. We can accomplish this with infrequent baked-in data updates, and we run little-to-no risk of redirecting *non* AWS traffic to AWS as Amazon and other providers are not in the habit of selling off their valuable IP addresses (only adding new ones infrequently). And we will not risk runtime failures.

* we do also have some resources from packet/equinix, but not exactly artifact hosting. Azure is doing some testing, but not provided through SIG K8s Infra (there's a stale issue in k/k8s.io about donating accounts to K8s Infra that has not moved).

**Comment by hh** [comment]
I agree let's laser focus on AWS.

Let's start with ip-ranges.json, but understand that we may need to include AWS ASN IP ranges later.

Our current spend calculations have shown include the ASN data, and I am under the impression it may be a considerable percentage of the money. @Riaankl ?

**Comment by Riaankl** [comment]
To generate the Data Studio Graphs for consumption per provider we used not only the content of [ip-ranges.json](https://ip-ranges.amazonaws.com/ip-ranges.json) but also [ANS data](https://github.com/kubernetes/k8s.io/blob/main/registry.k8s.io/infra/meta/asns/amazon.yaml) in attempt to assign the largest possible number of IPs to their respective owners.

We are busy splitting out the differance between `Amazon.ip-ranges.json` and `Amazon.ASN` to see the remainder traffic that would not be caught by redirecting based on  [ip-ranges.json](https://ip-ranges.amazonaws.com/ip-ranges.json)
Once we have the data, will share the graph showing the split.

**Comment by Riaankl** [comment]
We updated the log data to see the differance between Amazon.ip-ranges.json and Amazon.ASN.
Our conclusion is the for the last 30 days 6.5% of the traffic belong to Amazon IPs that was identified using the ASN and not the  ip-ranges.json.

**Comment by hh** [comment]
Thanks for putting this in the backlog.

**Comment by dims** [comment]
@BenTheElder @BobyMCbobs So in OCI-proxy, we use something like https://github.com/yl2chen/cidranger with the data in https://ip-ranges.amazonaws.com/ip-ranges.json and we get a thumbs up/down if the incoming IP address is from AWS or not (and we would route the traffic to cloudfront at that point if so)?

**Comment by BenTheElder** [comment]
> Ted and I are currently calculating a rough amount of costs for an S3-only versus CloudFront+S3 solution and seeing whether the additional costs of a CloudFront-based solution (which of course would be more expensive since it's a managed solution) would be worth it. In other words, if oci-proxy only needs to determine IF a source IP is coming from AWS (and not which AWS region the IP is in), that would reduce the amount of computing needed in the Cloud Run container and offload that to CloudFront...

I'm not really worried about the compute in the redirector, the cost will be pretty similar to detect region or detect AWS on the whole we can use fast matching for both.

What I do want to know is what is most reasonable on the AWS side for our problem, so we're not wasting unnecessary resources there. On the GCP side we've wound up doing the equivalent of regional redirecting in a special service ~ the registry. If cloudfront makes more sense, great!

> @BenTheElder @BobyMCbobs So in OCI-proxy, we use something like https://github.com/yl2chen/cidranger with the data in https://ip-ranges.amazonaws.com/ip-ranges.json and we get a thumbs up/down if the incoming IP address is from AWS or not (and we would route the traffic to cloudfront at that point if so)?

I have code with benchmarks including exactly this package, but I'm not really concerned with rushing to merge this! We should evaluate the best plan while we continue to evaluate the best option on the AWS side and actually get the AWS resources and image promotion in place, and in parallel continue to roll out the simple redirect service.

To support rolling out the simple redirect service, I'd actually like to *not* add any further complexity to the binary until we have a tagged release with just redirecting + necessary future proofing (think https://github.com/kubernetes-sigs/oci-proxy/pull/28) so we have a safe place to roll back to in the deployment and we can rapidly roll out the production domain.

I fear if we don't get the domain rolled out sooner rather than later, the rest is moot.

**Comment by BenTheElder** [comment]
Given that going forward the proxy is responsible for routing to regionalized copies, we definitely need data more specific than the organization (specifically we need to know which AWS region is associated with the client IP).

We've determined with AWS that this (s3 bucket per region we host, redirect to those buckets based on client location) is the cost effective route to go architecturally given our constraints on hosting a public-read blob-only-mirror. Serving clients with a copy in their own region is extremely cheap, latency and costs are not cheap otherwise.

Unless we switch approaches entirely to one that still involves AWS hosting. a mirror but with a global auto-regionalizing endpoint we're not responsible for (ECR?), we won't be using ASN data.

Thanks all.
